### PR TITLE
*: support set relation job

### PR DESCRIPTION
**Pull request by zimulala**
We will initially support the parallel DDL. We need to wait for PR(#6161) to merge before we can begin.
In order to prevent conflicts, I first implement the function of `setJobRelation`.
setJobRelation sets the current job's relation-ID to the ID of a job that is related to the current job.

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @winkyao @coocood @shenli

**Comment by coocood** [comment]
Have you considered `DependsOn` instead of `RelatedTo`?

File: ddl/ddl_worker.go [hunk 1]
```diff
@@ -91,6 +91,29 @@ func (d *ddl) isOwner() bool {
...
+func setJobRelation(t *meta.Meta, curJob *model.Job) error {
+	jobs, err := t.GetAllDDLJobs()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, job := range jobs {
+		if curJob.ID < job.ID {
+			continue
+		}
+		isRelated, err := curJob.IsRelatedJob(job)
```
Comment (zhexuany) [inline ddl/ddl_worker.go]:
```
if isRelated, err := curJob.IsRelatedJob(job); err == nil; {
          if isRelated {
       		curJob.RelatedID = job.ID
  			break
 		}
} else {
       return err
}
```
above way is more concise?

File: ddl/ddl_worker.go [hunk 2]
```diff
@@ -91,6 +91,29 @@ func (d *ddl) isOwner() bool {
 	return isOwner
 }
 
+// setJobRelation sets the current job's relation-ID to the ID of a job that is related to current job.
+// The related job's ID must less than the current job's ID, and we need the largest one in the list.
+func setJobRelation(t *meta.Meta, curJob *model.Job) error {
```
[the remark this answers is not in the document]
Comment (winkyao) [reply ddl/ddl_worker.go]: Does setJobRelation run in a transaction?

File: model/ddl.go [hunk 3]
```diff
@@ -213,6 +215,41 @@ func (job *Job) String() string {
 		job.ID, job.Type, job.State, job.SchemaState, job.SchemaID, job.TableID, rowCount, len(job.Args), tsConvert2Time(job.StartTS), job.Error, job.ErrorCount, job.SnapshotVer)
 }
 
+func (job *Job) isRelatedSchema(job1 *Job) (bool, error) {
+	if job1.Type == ActionDropSchema || job1.Type == ActionCreateSchema {
+		if job1.SchemaID == job.SchemaID {
+			return true, nil
+		}
+		if job.Type == ActionRenameTable {
+			var schemaID int64
```
Comment (winkyao) [inline model/ddl.go]: Name it `oldSchemaID` is more clear?

File: model/ddl.go [hunk 4]
```diff
@@ -213,6 +215,41 @@ func (job *Job) String() string {
...
+			if job1.SchemaID == schemaID {
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+// IsRelatedJob returns whether job and job1 are related or not.
+func (job *Job) IsRelatedJob(job1 *Job) (bool, error) {
```
Comment (winkyao) [inline model/ddl.go]: job1 ---> other

File: model/ddl.go [hunk 5]
```diff
@@ -213,6 +215,41 @@ func (job *Job) String() string {
 		job.ID, job.Type, job.State, job.SchemaState, job.SchemaID, job.TableID, rowCount, len(job.Args), tsConvert2Time(job.StartTS), job.Error, job.ErrorCount, job.SnapshotVer)
 }
 
+func (job *Job) isRelatedSchema(job1 *Job) (bool, error) {
```
Comment (winkyao) [inline model/ddl.go]: rename job1 to `other`?

File: structure/list.go [hunk 6]
```diff
@@ -136,6 +136,26 @@ func (t *TxStructure) LLen(key []byte) (int64, error) {
...
+func (t *TxStructure) LGetAll(key []byte) ([][]byte, error) {
+	metaKey := t.encodeListMetaKey(key)
+	meta, err := t.loadListMeta(metaKey)
+	if err != nil || meta.IsEmpty() {
+		return nil, errors.Trace(err)
+	}
+
+	length := int(meta.RIndex - meta.LIndex)
+	elements := make([][]byte, 0, length)
+	for index := meta.RIndex - 1; index >= meta.LIndex; index-- {
```
Comment (ciscoxll) [inline structure/list.go]: Can you be sure it's ordered.

File: meta/meta_test.go [hunk 7]
```diff
@@ -1,3 +1,4 @@
+// LGetAll gets all elements of this list in order from left to right.
```
Comment (lamxTyler) [inline meta/meta_test.go]: Why put the comment here?

**Comment by zimulala** [comment]
PTAL @coocood @winkyao @zhexuany @lamxTyler @ciscoxll

**Comment by zimulala** [comment]
PTAL @coocood @winkyao

File: ddl/ddl_worker.go [hunk 8]
```diff
@@ -91,6 +91,29 @@ func (d *ddl) isOwner() bool {
 	return isOwner
 }
 
+// buildJobDependence sets the current job's dependence-ID that the current job depends on.
+// The dependent job's ID must less than the current job's ID, and we need the largest one in the list.
+func buildJobDependence(t *meta.Meta, curJob *model.Job) error {
```
Comment (shenli) [inline ddl/ddl_worker.go]: Add comments for what is a dependence job?

File: model/ddl.go [hunk 9]
```diff
@@ -127,6 +127,8 @@ type Job struct {
 	// StartTS uses timestamp allocated by TSO.
 	// Now it's the TS when we put the job to TiKV queue.
 	StartTS uint64 `json:"start_ts"`
+	// DependentID is the job's ID that the current job depends on.
+	DependentID int64 `json:"related_id"`
```
Comment (coocood) [inline model/ddl.go]: s/DependentID/DependencyID
s/related_id/dependency_id

**Comment by zimulala** [comment]
PTAL @coocood @shenli

File: ddl/ddl_worker.go [hunk 10]
```diff
@@ -91,6 +91,29 @@ func (d *ddl) isOwner() bool {
 	return isOwner
 }
 
+// buildJobDependence sets the curjob's dependency-ID.
+// The dependency-job's ID must less than the current job's ID, and we need the largest one in the list.
+func buildJobDependence(t *meta.Meta, curJob *model.Job) error {
+	jobs, err := t.GetAllDDLJobs()
```
Comment (shenli) [inline ddl/ddl_worker.go]: Should we run this check in a transaction? For example, two DDLs come in the same time. Will this function detect the dependence?

Comment (zimulala) [reply ddl/ddl_worker.go]: It's in a transaction. The `t` has a field of transaction.

**Comment by zhexuany** [comment]
LGTM

**Pull request merged**